gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1353 @@
+				ereport(ERROR,
+						(errcode(ERRCODE_GP_COMMAND_ERROR),
+						 errmsg("database %u is not tracked", MyDatabaseId),
+					errhint("Call 'arenadata_toolkit.tracking_register_db()'"
+							"to enable tracking")));
+		}
+		else
+		{
+			/*
+			 * This code is needed for the cases when there are several track
+			 * requests within the same transaction. rollback_bloom stands for
+			 * preserving initial filter state at the moment of the first
+			 * function call within the transaction.
+			 */
+			if (tf_get_global_state.rollback_bloom == NULL)
+			{
+				tf_get_global_state.rollback_bloom = palloc(FULL_BLOOM_SIZE(bloom_size));
+				bloom_copy(tf_get_global_state.bloom, tf_get_global_state.rollback_bloom);
+			}
+			bloom_clear(tf_get_global_state.bloom);
+			if (!bloom_set_move(&tf_shared_state->bloom_set, MyDatabaseId, tf_get_global_state.bloom))
+				ereport(ERROR,
+						(errcode(ERRCODE_GP_COMMAND_ERROR),
+						 errmsg("database %u is not tracked", MyDatabaseId),
+					errhint("Call 'arenadata_toolkit.tracking_register_db()'"
+							"to enable tracking")));
+		}
+		/* initial snapshot shouldn't return drops */
+		if (tf_get_global_state.bloom && !tf_get_global_state.bloom->is_set_all)
+		{
+			tf_get_global_state.drops = drops_track_move(MyDatabaseId);
+			tf_get_global_state.next_drop = list_head(tf_get_global_state.drops);
+		}
+
+		/*
+		 * Let's retrieve tracking information only once for the transaction.
+		 */
+		if (tf_get_global_state.schema_oids == NIL)
+			get_filters_from_guc();
+
+		if (tf_get_global_state.relstorages == 0 ||
+			tf_get_global_state.relkinds == 0 ||
+			tf_get_global_state.schema_oids == NIL)
+			ereport(ERROR,
+					(errmsg("Cannot get tracking configuration (schemas, relkinds, reltorage) for database %u", MyDatabaseId)));
+
+		MemoryContextSwitchTo(oldcontext);
+
+		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
+
+		funcctx->tuple_desc = CreateTemplateTupleDesc(GET_TRACK_TUPDESC_LEN, false);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 1, "relid", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 2, "name", NAMEOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 3, "relfilenode", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 4, "size", INT8OID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 5, "state", CHAROID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 6, "gp_segment_id", INT4OID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 7, "relnamespace", OIDOID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 8, "relkind", CHAROID, -1, 0);
+		TupleDescInitEntry(funcctx->tuple_desc, (AttrNumber) 9, "relstorage", CHAROID, -1, 0);
+		funcctx->tuple_desc = BlessTupleDesc(funcctx->tuple_desc);
+
+		state = (tf_main_func_state_t *) palloc0(sizeof(tf_main_func_state_t));
+		funcctx->user_fctx = (void *) state;
+
+		if (tf_get_global_state.bloom)
```

Comment:
Is it possible for `tf_get_global_state.bloom` to be NULL here? Do we need the NULL check?